src/category_theory/groupoid/subgroupoid.lean

```diff
@@ -0,0 +1,356 @@
+/-
+Copyright (c) 2022 Rémi Bottinelli. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémi Bottinelli, Junyan Xu
+-/
+import category_theory.groupoid.vertex_group
+import category_theory.groupoid
+import algebra.group.defs
+import algebra.hom.group
+import algebra.hom.equiv
+import data.set.lattice
+import combinatorics.quiver.connected_component
+import group_theory.subgroup.basic
+/-!
+# Subgroupoid
+
+This file defines subgroupoids as `structure`s containing the subsets of arrows and their
+stability under composition and inversion.
+Also defined are
+
+* containment of subgroupoids is a complete lattice;
+* images and preimages of subgroupoids under a functor;
+* the notion of normality of subgroupoids and its stability under intersection and preimage;
+* compatibility of the above with `groupoid.vertex_group`.
+
+
+## Main definitions
+
+Given a type `C` with associated `groupoid C` instance.
+
+* `subgroupoid C` is the type of subgroupoids of `C`
+* `subgroupoid.is_normal` is the property that the subgroupoid is stable under conjugation
+  by arbitrary arrows, _and_ that all identity arrows are contained in the subgroupoid.
+* `subgroupoid.comap` is the "preimage" map of subgroupoids along a functor.
+* `subgroupoid.map` is the "image" map of subgroupoids along a functor _injective on objects_.
+* `subgroupoid.vertex_subgroup` is the subgroup of the `vertex group` at a given vertex `v`,
+  assuming `v` is contained in the `subgroupoid` (meaning, by definition, that the arrow `𝟙 v`
+  is contained in the subgroupoid).
+
+## Implementation details
+
+The structure of this file is copied from/inspired by `group_theory.subgroup.basic`
+and `combinatorics.simple_graph.subgraph`.
+
+## TODO
+
+* Equivalent inductive characterization of generated (normal) subgroupoids.
+* Characterization of normal subgroupoids as kernels.
+
+## Tags
+
+subgroupoid
+
+-/
+
+open set classical function
+
+namespace category_theory
+
+open groupoid
+
+local attribute [protected] category_theory.inv
+
+universes u v
+
+variables {C : Type u} [groupoid C]
+
+/--
+A sugroupoid of `C` consists of a choice of arrows for each pair of vertices, closed
+under composition and inverses
```

Comment:
```suggestion
under composition and inverses.
```
I think the general rule is to add a period if the docstring is a complete sentence.